File: multi-gammon-util/src/main/java/org/akoshterek/backgammon/train/NetworkHolder.java
package org.akoshterek.backgammon.train;

import org.akoshterek.backgammon.board.PositionClass;
import org.apache.commons.lang3.SerializationUtils;
import org.encog.neural.networks.BasicNetwork;
import org.encog.neural.networks.training.propagation.TrainingContinuation;
import org.encog.persist.EncogDirectoryPersistence;

import java.io.*;

/**
 * @author Alex
 *         date 22.09.2015.
 */
class NetworkHolder implements Serializable {
    private static final String DIRECTORY = "bin/";
    private static final long serialVersionUID = -7252112052540032945L;

    NetworkHolder(final BasicNetwork network, final PositionClass networkType) {
        this.network = network;
        this.networkType = networkType;
    }

    private BasicNetwork network;
    private int epoch = 1;
    private PositionClass networkType;
    private TrainingContinuation continuation;

    void serialize(final AgentSettings agentSettings) {
        File file = new File(getResumeFileName(agentSettings, networkType));
        if(!file.getParentFile().exists()) {
            file.getParentFile().mkdirs();
        }
        try(OutputStream os = new FileOutputStream(file)) {
            SerializationUtils.serialize(this, os);
        } catch (final Exception e) {
            throw new RuntimeException(e);
        }
    }

    void serializeTrainedNetwork(final AgentSettings agentSettings) {
        File file = new File(getTrainedNetworkFileName(agentSettings, networkType));
        if(!file.getParentFile().exists()) {
            file.getParentFile().mkdirs();
        }

        EncogDirectoryPersistence.saveObject(file, network);
    }

    static BasicNetwork deserializeTrainedNetwork(final AgentSettings agentSettings, PositionClass networkType) {
        File file = new File(getTrainedNetworkFileName(agentSettings, networkType));
        if(file.exists()) {
            return (BasicNetwork) EncogDirectoryPersistence.loadObject(file);
        } else {
            return null;
        }
    }

    public static NetworkHolder deserialize(NetworkHolder template, AgentSettings agentSettings) {
        try(InputStream is = new FileInputStream(getResumeFileName(agentSettings, template.networkType))) {
            return SerializationUtils.<NetworkHolder>deserialize(is);
        } catch (FileNotFoundException e) {
            return null;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public BasicNetwork getNetwork() {
        return network;
    }

    public int getEpoch() {
        return epoch;
    }

    public void incEpoch() {
        epoch++;
    }

    private static String getResumeFileName(AgentSettings agentSettings, PositionClass networkType) {
        return DIRECTORY + agentSettings.agentName() + "-" + PositionClass.getNetworkType(networkType) + "-resume.egs";
    }

    private static String getTrainedNetworkFileName(AgentSettings agentSettings, PositionClass networkType) {
        return DIRECTORY + agentSettings.agentName() + "-" + PositionClass.getNetworkType(networkType) + ".eg";
    }

    public TrainingContinuation getContinuation() {
        return continuation;
    }

    public void setContinuation(TrainingContinuation continuation) {
        this.continuation = continuation;
    }
}

File: multi-gammon-util/src/main/java/org/akoshterek/backgammon/train/NetworkHolder.scala
package org.akoshterek.backgammon.train

import java.io._

import org.akoshterek.backgammon.board.PositionClass
import org.apache.commons.lang3.SerializationUtils
import org.encog.neural.networks.BasicNetwork
import org.encog.neural.networks.training.propagation.TrainingContinuation
import org.encog.persist.EncogDirectoryPersistence
import resource.managed

/**
  * @author Alex
  *         date 22.09.2015.
  */
@SerialVersionUID(-7252112052540032945L)
object NetworkHolder {
  private val DIRECTORY: String = "bin/"

  def deserializeTrainedNetwork(agentSettings: AgentSettings, networkType: PositionClass): Option[BasicNetwork] = {
    val file: File = new File(getTrainedNetworkFileName(agentSettings, networkType))
    if (file.exists) {
      Some(EncogDirectoryPersistence.loadObject(file).asInstanceOf[BasicNetwork])
    }
    else {
      None
    }
  }

  def deserialize(template: NetworkHolder, agentSettings: AgentSettings): Option[NetworkHolder] = {
    val file = new File(getResumeFileName(agentSettings, template.networkType))
    if (!file.exists()) {
      None
    } else {
      Some(
        managed(new FileInputStream(file)).acquireFor(is => {
          SerializationUtils.deserialize[NetworkHolder](is)
        }).either.right.get
      )
    }
  }

  private def getResumeFileName(agentSettings: AgentSettings, networkType: PositionClass): String = {
    DIRECTORY + agentSettings.agentName + "-" + PositionClass.getNetworkType(networkType) + "-resume.egs"
  }

  private def getTrainedNetworkFileName(agentSettings: AgentSettings, networkType: PositionClass): String = {
    DIRECTORY + agentSettings.agentName + "-" + PositionClass.getNetworkType(networkType) + ".eg"
  }
}

@SerialVersionUID(-7252112052540032945L)
class NetworkHolder private[train](val network: BasicNetwork, val networkType: PositionClass) extends Serializable {
  private var _epoch: Int = 1
  private var _continuation: TrainingContinuation = _

  def serialize(agentSettings: AgentSettings): Unit = {
    val file: File = new File(NetworkHolder.getResumeFileName(agentSettings, networkType))
    if (!file.getParentFile.exists) {
      file.getParentFile.mkdirs
    }

    managed(new FileOutputStream(file)).acquireAndGet(os => {
      SerializationUtils.serialize(this, os)
    })
  }

  private[train] def serializeTrainedNetwork(agentSettings: AgentSettings) {
    val file: File = new File(NetworkHolder.getTrainedNetworkFileName(agentSettings, networkType))
    if (!file.getParentFile.exists) {
      file.getParentFile.mkdirs
    }
    EncogDirectoryPersistence.saveObject(file, network)
  }

  def epoch: Int = _epoch

  def incEpoch(): Unit = _epoch += 1

  def continuation: TrainingContinuation = _continuation
  def continuation_=(cont: TrainingContinuation): Unit = _continuation = cont
}